File: feed-wrangler/healthcheck.js
#!/usr/bin/env node
// Healthcheck for signal-clearance (distroless/nodejs container).
// Raw RESP over net sockets — zero npm dependencies.
// Watchdog Redis pub/sub mechanism, same as the Go healthcheck binary.
'use strict';

const net = require('net');
const crypto = require('crypto');

const redisAddr  = (process.env.REDIS_URL || 'redis:6379').split(':');
const redisHost  = redisAddr[0] || 'redis';
const redisPort  = parseInt(redisAddr[1] || '6379', 10);
const svcName    = process.env.SERVICE_NAME || 'signal-clearance';
const timeoutMs  = parseInt(process.env.CHECK_TIMEOUT_MS || '8000', 10);

const requestId     = `chk-${crypto.randomBytes(8).toString('hex')}`;
const resultChannel = `tca:check-results:${requestId}`;

let done = false;
const exit = (code) => {
  if (done) return;
  done = true;
  setTimeout(() => process.exit(code), 50);
};

const overallTimeout = setTimeout(() => {
  // Watchdog did not respond in time — treat as degraded-healthy.
  // The Job is running; the Watchdog will do behavioural verification
  // on the next cycle.
  console.error('[healthcheck] Timeout — degraded-healthy');
  exit(0);
}, timeoutMs);

// ---------------------------------------------------------------------------
// RESP command builder
// ---------------------------------------------------------------------------
function resp(...args) {
  let out = `*${args.length}\r\n`;
  for (const arg of args) {
    const s = String(arg);
    out += `$${Buffer.byteLength(s, 'utf8')}\r\n${s}\r\n`;
  }
  return out;
}

// ---------------------------------------------------------------------------
// Scan accumulated buffer for a complete JSON object with a healthy field.
// Returns true (healthy), false (unhealthy), or null (not yet available).
// ---------------------------------------------------------------------------
function scanForResult(buf) {
  let i = 0;
  while (i < buf.length) {
    const start = buf.indexOf('{', i);
    if (start === -1) break;

    let depth = 0, end = -1;
    for (let j = start; j < buf.length; j++) {
      if (buf[j] === '{') depth++;
      else if (buf[j] === '}') { if (--depth === 0) { end = j; break; } }
    }
    if (end === -1) break; // Incomplete — wait for more data

    try {
      const obj = JSON.parse(buf.slice(start, end + 1));
      if (typeof obj.healthy === 'number') {
        return obj.healthy === 1 ? true : false;
      }
    } catch (_) {}

    i = end + 1;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Subscriber connection — accumulate all data into a single buffer
// ---------------------------------------------------------------------------
let buffer = '';

const sub = net.createConnection({ host: redisHost, port: redisPort });
sub.setEncoding('utf8');

sub.on('error', (err) => {
  console.error(`[healthcheck] sub error: ${err.message} — degraded-healthy`);
  clearTimeout(overallTimeout);
  exit(0);
});

sub.on('connect', () => {
  sub.write(resp('SUBSCRIBE', resultChannel));
  // Wait 100 ms to ensure Redis has registered the subscription
  // before we publish the check request on the publisher connection.
  setTimeout(openPublisher, 100);
});

sub.on('data', (chunk) => {
  buffer += chunk;
  const result = scanForResult(buffer);
  if (result === null) return; // Still waiting
  clearTimeout(overallTimeout);
  if (result) {
    exit(0);
  } else {
    console.error('[healthcheck] Watchdog returned unhealthy');
    exit(1);
  }
});

// ---------------------------------------------------------------------------
// Publisher connection — send check request, then destroy
// ---------------------------------------------------------------------------
function openPublisher() {
  const payload = JSON.stringify({
    request_id:   requestId,
    container_id: process.env.HOSTNAME || 'unknown',
    service_name: svcName,
    published_at: new Date().toISOString(),
  });

  const pub = net.createConnection({ host: redisHost, port: redisPort });
  pub.setEncoding('utf8');

  pub.on('error', (err) => {
    console.error(`[healthcheck] pub error: ${err.message} — degraded-healthy`);
    clearTimeout(overallTimeout);
    exit(0);
  });

  pub.on('connect', () => {
    pub.write(resp('PUBLISH', 'tca:check-requests', payload));
    setTimeout(() => { try { pub.destroy(); } catch (_) {} }, 300);
  });
}

File: results/healthcheck.py
#!/usr/bin/env python3
"""
Healthcheck for results (distroless/python3 container).
Raw RESP protocol over socket — zero external dependencies.
Same Watchdog Redis pub/sub mechanism as Go healthcheck binary.
"""

import os
import socket
import time
import secrets
import json
import sys

REDIS_ADDR = os.environ.get('REDIS_URL', 'redis:6379').split(':')
REDIS_HOST = REDIS_ADDR[0]
REDIS_PORT = int(REDIS_ADDR[1]) if len(REDIS_ADDR) > 1 else 6379
SERVICE_NAME = os.environ.get('SERVICE_NAME', 'results')
TIMEOUT_MS = int(os.environ.get('CHECK_TIMEOUT_MS', '8000'))
CONTAINER_ID = os.environ.get('HOSTNAME', 'unknown')

request_id = f'chk-{secrets.token_hex(8)}'
result_channel = f'tca:check-results:{request_id}'

def resp(*args):
    """Build a RESP command."""
    out = f'*{len(args)}\r\n'
    for arg in args:
        s = str(arg)
        out += f'${len(s.encode())}\r\n{s}\r\n'
    return out.encode()

def exit_degraded_healthy(reason):
    print(f'[healthcheck] {reason} — degraded-healthy', file=sys.stderr)
    sys.exit(0)

def exit_unhealthy(reason):
    print(f'[healthcheck] {reason}', file=sys.stderr)
    sys.exit(1)

try:
    # Subscriber connection — subscribe before publishing to avoid race
    sub_sock = socket.create_connection((REDIS_HOST, REDIS_PORT), timeout=3)
    sub_sock.sendall(resp('SUBSCRIBE', result_channel))
    # Read subscribe confirmation
    sub_sock.recv(256)

    # Publisher connection — send check request then close
    pub_sock = socket.create_connection((REDIS_HOST, REDIS_PORT), timeout=3)
    payload = json.dumps({
        'request_id': request_id,
        'container_id': CONTAINER_ID,
        'service_name': SERVICE_NAME,
        'published_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
    })
    pub_sock.sendall(resp('PUBLISH', 'tca:check-requests', payload))
    pub_sock.recv(64)
    pub_sock.close()

    # Wait for Watchdog response
    deadline = time.time() + (TIMEOUT_MS / 1000)
    sub_sock.settimeout(TIMEOUT_MS / 1000)
    buf = b''

    while time.time() < deadline:
        try:
            chunk = sub_sock.recv(4096)
            if not chunk:
                break
            buf += chunk

            # Find JSON object in buffer
            start = buf.find(b'{')
            if start == -1:
                continue

            depth = 0
            end = -1
            for i in range(start, len(buf)):
                if buf[i:i+1] == b'{':
                    depth += 1
                elif buf[i:i+1] == b'}':
                    depth -= 1
                    if depth == 0:
                        end = i
                        break

            if end == -1:
                continue

            result = json.loads(buf[start:end+1])
            sub_sock.close()

            if result.get('healthy') == 1:
                sys.exit(0)
            else:
                exit_unhealthy(f"Unhealthy: {result.get('failure_reason', 'unknown')}")

        except socket.timeout:
            break

    sub_sock.close()
    exit_degraded_healthy('Timeout waiting for Watchdog response')

except OSError as e:
    exit_degraded_healthy(f'Redis connection error: {e}')
